rustup-init.ps1

```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env powershell
+
+# This is just a little script that can be downloaded from the internet to
+# install rustup. It just does platform detection, downloads the installer and
+# runs it.
+
+#XXX: If you change anything here, please make the same changes in setup_mode.rs
+function Write-Usage {
+    Write-Host @'
+rustup-init 1.18.3 (435397f48 2019-05-22)
+The installer for rustup
+
+USAGE:
+    rustup-init.exe [FLAGS] [OPTIONS]
+
+FLAGS:
+    -v, --verbose           Enable verbose output
+    -y                      Disable confirmation prompt.
+        --no-modify-path    Don't configure the PATH environment variable
+    -h, --help              Prints help information
+    -V, --version           Prints version information
+
+OPTIONS:
+        --default-host <default-host>              Choose a default host triple
+        --default-toolchain <default-toolchain>    Choose a default toolchain to install
+'@
+}
+
+function main([AllowEmptyCollection()] [string[]]$Argv = @()) {
+    foreach ($arg in $Argv) {
+        if ("$arg" -eq "-h" -or "$arg" -eq "--help") {
+            Write-Usage
+            return
+        }
+    }
+
+    Install-Rustup $Argv
+}
+
+function Get-Architecture {
+    $platform = [System.Environment]::OSVersion.Platform
+    switch -Wildcard ($platform) {
+        # Linux and Darwin both report as Unix
+        "Unix" {
+            $ostype = uname -s
+            $cputype = uname -m
+            break
+        }
+        "Win*" {
+            $ostype = (Get-WmiObject Win32_OperatingSystem).Caption
+            $cputype = (Get-WmiObject Win32_OperatingSystem).OSArchitecture
+            break
+        }
+        default {
+            Write-Error "could not determine OS platform: $platform"
+        }
+    }
+
+    switch -Wildcard ($ostype) {
+        "Darwin" {
+            $ostype = "apple-darwin"
+            break
+        }
+        "Linux" {
+            $ostype = "unknown-linux-gnu"
+            break
+        }
+        "*Windows*" {
+            $ostype = "pc-windows-msvc"
+            break
+        }
+        default {
+            Write-Error "unrecognized OS type: $ostype"
+        }
+    }
+
+    switch ($cputype) {
+        # Reported by Windows 32-bit operating systems
+        "32-bit" {
+            $cputype = "i686"
+            break
+        }
+        # Reported by Windows 64-bit operating systems
+        "64-bit" {
+            $cputype = "x86_64"
+            break
+        }
+        # Reported by Darin/Linux 64-bit operating systems
+        "x86_64" {
+            break
+        }
+        default {
+            Write-Error "unknown CPU type: $cputype"
+        }
+    }
+
+    $arch = "${cputype}-${ostype}"
+    Write-Debug "Detected arch '$arch'"
+    return "$arch"
+}
+
+# If RUSTUP_UPDATE_ROOT is unset, default it.
+function Get-RustupUpdateRoot {
+    if (Test-Path env:\RUSTUP_UPDATE_ROOT) {
```

Comment:
is the backslash intended here?

Reply:
It turns out, that, yes, it is intended here. In PowerShell filesystems, but also the environment variables are contained in PowerShell Drives, and the `Env:` is no different:

```powershell
> Get-PSDrive

Name           Used (GB)     Free (GB) Provider      Root                                                                                                                                                                                                                                                             CurrentLocation
----           ---------     --------- --------      ----                                                                                                                                                                                                                                                             ---------------
Alias                                  Alias
C                   0.98         18.90 FileSystem    C:\                                                                                                                                                                                                                           Users\fnichol\Projects\github.com\fnichol\mtoc\.ci
Cert                                   Certificate   \
Env                                    Environment
Function                               Function
HKCU                                   Registry      HKEY_CURRENT_USER
HKLM                                   Registry      HKEY_LOCAL_MACHINE
Variable                               Variable
WSMan                                  WSMan
```
In this case the `Test-Path` cmdlet works on any PowerShell drives, hence that backslash that looks like a mistake here, but for the `C` drive, it would be `C:\temp`. I know, I've learned some things here too

Reply:
Wow, that's fun.  Thanks for the explanation.

Reply:
nit: To be more correct. The ENV: drive is a PowerShell Provider. And you traverse providers like a filesystem.  Think inbuilt-Fuse.

Reply:
Ah, you are totally right and I forgot to qualify that